### PR TITLE
Allow live pdf template edit

### PR DESCRIPTION
**Pull request by Roshyo**
Related to #187 

image

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -70,7 +70,13 @@
         ]
     },
     "config": {
-        "sort-packages": true
+        "sort-packages": true,
+        "allow-plugins": {
+            "composer/package-versions-deprecated": true,
+            "dealerdirect/phpcodesniffer-composer-installer": true,
```
Comment (loevgaard) [inline composer.json]: We don't need this to run I think

File: src/Controller/Action/AdminGeneratePdfAction.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+use Setono\SyliusGiftCardPlugin\Factory\GiftCardFactoryInterface;
+use Setono\SyliusGiftCardPlugin\Form\Type\GiftCardConfigurationType;
+use Setono\SyliusGiftCardPlugin\Generator\GiftCardPdfGeneratorInterface;
+use Setono\SyliusGiftCardPlugin\Model\GiftCardConfigurationInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+final class AdminGeneratePdfAction
```
Comment (loevgaard) [inline src/Controller/Action/AdminGeneratePdfAction.php]: Put inside `Admin` folder instead?

File: src/Factory/GiftCardFactoryInterface.php [hunk 3]
```diff
@@ -24,4 +24,6 @@ public function createFromOrderItemUnitAndCart(
         OrderItemUnitInterface $orderItemUnit,
         OrderInterface $cart
     ): GiftCardInterface;
+
+    public function createDummy(): GiftCardInterface;
```
Comment (loevgaard) [inline src/Factory/GiftCardFactoryInterface.php]: What is this used for? :) Could we have a better name?

File: src/Generator/GiftCardPdfGenerator.php [hunk 4]
```diff
@@ -33,4 +39,21 @@ public function generatePdfResponse(
 
         return new PdfResponse($this->snappy->getOutputFromHtml($html), 'gift_card.pdf');
     }
+
+    public function generateAndSavePdf(
```
Comment (loevgaard) [inline src/Generator/GiftCardPdfGenerator.php:68]: Why isn't this method returning anything? :)

Comment (Roshyo) [reply src/Generator/GiftCardPdfGenerator.php:68]: It is just generating the PDF and creating the file. So nothing to return here imo.

Comment (loevgaard) [reply src/Generator/GiftCardPdfGenerator.php:68]: But wouldn't you expect from the name of the method that you either give the filename as input or it returns the filename as output?

Comment (Roshyo) [reply src/Generator/GiftCardPdfGenerator.php:68]: Well, Since the file path generation is handled in another service, I'm not sure it is necessary. But we can return the name indeed.

File: src/Controller/Action/Admin/RenderPdfAction.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        RepositoryInterface $giftCardConfigurationRepository,
+        GiftCardPdfPathGeneratorInterface $giftCardPdfPathGenerator,
+        Filesystem $filesystem
+    ) {
+        $this->giftCardConfigurationRepository = $giftCardConfigurationRepository;
+        $this->giftCardPdfPathGenerator = $giftCardPdfPathGenerator;
+        $this->filesystem = $filesystem;
+    }
+
+    public function __invoke(Request $request, int $id): PdfResponse
```
Comment (loevgaard) [inline src/Controller/Action/Admin/RenderPdfAction.php]: I would have expected that the `$id` was the id of the gift card, but it's not apparently. Can you explain the flow of this. I don't seem to get it

File: src/Factory/DummyGiftCardFactory.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Setono\SyliusGiftCardPlugin\Factory;
+
+use Setono\SyliusGiftCardPlugin\Model\GiftCardInterface;
+use Sylius\Component\Currency\Context\CurrencyContextInterface;
+
+final class DummyGiftCardFactory implements DummyGiftCardFactoryInterface
```
Comment (loevgaard) [inline src/Factory/DummyGiftCardFactory.php]: Can you explain the usage of this 'DummyGiftCard' ?

File: src/Resources/config/doctrine/model/GiftCardConfiguration.orm.xml [hunk 7]
```diff
@@ -14,6 +14,7 @@
             </options>
         </field>
         <field name="defaultValidityPeriod" nullable="true" />
+        <field name="pdfRenderingCss" type="text" nullable="true" />
```
Comment (loevgaard) [inline src/Resources/config/doctrine/model/GiftCardConfiguration.orm.xml:19]: What is this? Can I input CSS that will be applied to the gift card?

File: src/Controller/Action/Admin/RenderPdfAction.php (hunk 5 above)
Comment (Roshyo) [reply src/Controller/Action/Admin/RenderPdfAction.php]: Well, we are in the admin, configuring a GiftCardConfiguration. We don't have any gift card right now (hence the dummy factory) but we just want to render the PDF that will be given to any user buying a gift card with this configuration

File: src/Factory/DummyGiftCardFactory.php (hunk 6 above)
Comment (Roshyo) [reply src/Factory/DummyGiftCardFactory.php]: See above

File: src/Resources/config/doctrine/model/GiftCardConfiguration.orm.xml (hunk 7 above)
Comment (Roshyo) [reply src/Resources/config/doctrine/model/GiftCardConfiguration.orm.xml:19]: Yes, exactly. No linter on this right now though :D

File: src/Resources/views/Admin/GiftCardConfiguration/Update/_javascripts.html.twig [hunk 8]
```diff
@@ -0,0 +1,6 @@
+<script src="{{ asset('bundles/setonosyliusgiftcardplugin/setono-sylius-gift-card-live-pdf-rendering.js') }}"></script>
```
Comment (loevgaard) [inline src/Resources/views/Admin/GiftCardConfiguration/Update/_javascripts.html.twig:1]: What's the advantage of having a file instead of just having the JS right here?

Comment (Roshyo) [reply src/Resources/views/Admin/GiftCardConfiguration/Update/_javascripts.html.twig:1]: We should never put JS directly into HTML imo. It is not the best practice

File: src/Resources/views/Admin/GiftCardConfiguration/_form.html.twig [hunk 9]
```diff
@@ -23,3 +23,33 @@
...
+
+            <button class="ui button primary labeled icon js-ssgc-apply-pdf-changes"
+                    data-url="{{ path('setono_sylius_gift_card_admin_ajax_generate_pdf', {'id': resource.id}) }}">
+                <i class="icon check"></i>
+                {{ 'setono_sylius_gift_card.ui.apply'|trans }}
+            </button>
+        </div>
+        <div class="column js-ssgc-live-render-container" style="height: 900px;">
+            <embed id="js-ssgc-live-render-box"
+                   src="{{ url('setono_sylius_gift_card_admin_partial_render_pdf', {id: resource.id}) }}"
```
Comment (loevgaard) [inline src/Resources/views/Admin/GiftCardConfiguration/_form.html.twig]: Imo you should just fetch the PDF on page load so that you don't need two endpoints

File: src/Controller/Action/Admin/RenderPdfAction.php (hunk 5 above)
Comment (loevgaard) [reply src/Controller/Action/Admin/RenderPdfAction.php]: Of course, haha. Well, the name of this action could be better, i.e. `RenderExamplePdfAction`

File: src/Factory/DummyGiftCardFactory.php (hunk 6 above)
Comment (loevgaard) [reply src/Factory/DummyGiftCardFactory.php]: I think `Example` is a better name than `Dummy`. Dummy in my vocabulary really means something you shouldn't use in production

File: src/Resources/views/Admin/GiftCardConfiguration/_form.html.twig (hunk 9 above)
Comment (Roshyo) [reply src/Resources/views/Admin/GiftCardConfiguration/_form.html.twig]: There were issues with the embed thing and dynamically generating the PDF, hence this flow

File: src/Factory/DummyGiftCardFactory.php (hunk 6 above)
Comment (Roshyo) [reply src/Factory/DummyGiftCardFactory.php]: Here is the definition I use: `an object designed to resemble and serve as a substitute for the real or usual one.`

Which is clearly what it does :thinking: `Example` in my mind would read as the same as `CustomerExampleFactory` in fixtures, and i'm not sure it is what we intend

File: src/Resources/views/Shop/GiftCard/pdf.html.twig [hunk 10]
```diff
@@ -9,16 +9,7 @@
...
-                font-family: sans-serif;
-                font-size: 3rem;
-                background: transparent url("{{ path }}") no-repeat left bottom;
-                background-size: 100px;
-            }
-
-            img {
-                display: block;
-            }
+            {{ configuration.pdfRenderingCss|default('')|ssgc_replace_css_option(_context)|raw }}
```
Comment (loevgaard) [inline src/Resources/views/Shop/GiftCard/pdf.html.twig:12]: How do we use some kind of default value if the user hasn't added any CSS?

We are also still missing the output of the gift card HTML on the gift card configuration page so that users easily can see the selectors they can use in the CSS, right?

File: src/Factory/DummyGiftCardFactory.php (hunk 6 above)
Comment (loevgaard) [reply src/Factory/DummyGiftCardFactory.php]: Exactly, a dummy is not something that should be used for real :D But this example is _real_ :D

File: src/Resources/views/Shop/GiftCard/pdf.html.twig (hunk 10 above)
Comment (Roshyo) [reply src/Resources/views/Shop/GiftCard/pdf.html.twig:12]: Right, there is no default value as of now.
Neither display of the HTML.
I'll add that.

File: composer-require-checker.json [hunk 11]
```diff
@@ -77,6 +77,7 @@
     "Sylius\\Component\\Product\\Resolver\\ProductVariantResolverInterface",
     "Sylius\\Component\\Promotion\\Checker\\Rule\\RuleCheckerInterface",
     "Sylius\\Component\\Promotion\\Model\\PromotionSubjectInterface",
-    "Sylius\\Component\\User\\Model\\UserInterface"
+    "Sylius\\Component\\User\\Model\\UserInterface",
+    "Symfony\\Component\\DomCrawler\\Crawler"
```
Comment (loevgaard) [inline composer-require-checker.json]: Shouldn't this be added to composer.json?

File: src/Factory/ExampleGiftCardFactory.php [hunk 12]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Setono\SyliusGiftCardPlugin\Factory;
+
+use Setono\SyliusGiftCardPlugin\Model\GiftCardInterface;
+use Sylius\Component\Currency\Context\CurrencyContextInterface;
+
+final class ExampleGiftCardFactory implements ExampleGiftCardFactoryInterface
```
Comment (loevgaard) [inline src/Factory/ExampleGiftCardFactory.php]: Why don't we have a method called `createExample()` on the `GiftCardFactoryInterface`?

File: src/Generator/GiftCardPdfGenerator.php [hunk 13]
```diff
@@ -16,10 +17,20 @@ class GiftCardPdfGenerator implements GiftCardPdfGeneratorInterface
...
-    {
+    private GiftCardPdfPathGeneratorInterface $giftCardPdfPathGenerator;
+
+    private Filesystem $filesystem;
+
+    public function __construct(
+        Environment $twig,
+        GeneratorInterface $snappy,
+        GiftCardPdfPathGeneratorInterface $giftCardPdfPathGenerator,
+        Filesystem $filesystem
```
Comment (loevgaard) [inline src/Generator/GiftCardPdfGenerator.php]: Use interface?

File: src/Resources/config/routes/admin_partial.yaml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+setono_sylius_gift_card_admin_partial_render_pdf:
+    path: /gift-card-configurations/{id}/render-pdf
+    controller: setono_sylius_gift_card.controller.action.admin_render
+    methods: [GET, POST, PUT]
```
Comment (loevgaard) [inline src/Resources/config/routes/admin_partial.yaml]: GET, POST, PUT? Do we need all of those?

File: src/Resources/config/routes/admin_ajax.yaml [hunk 15]
```diff
@@ -24,3 +24,17 @@ setono_sylius_gift_card_admin_ajax_customer_by_email:
             repository:
                 method: findBy
                 arguments: [email: $email]
+
+setono_sylius_gift_card_admin_ajax_generate_pdf:
+    path: /gift-card-configurations/{id}/generate-pdf
+    methods: [ GET, POST, PUT ]
```
Comment (loevgaard) [inline src/Resources/config/routes/admin_ajax.yaml]: GET, POST, PUT? Do we need all of those?

File: src/Resources/private/default-gift-card-configuration-css.txt [hunk 16]
```diff
@@ -0,0 +1,10 @@
+body {
```
Comment (loevgaard) [inline src/Resources/private/default-gift-card-configuration-css.txt]: Why not name this file `.css`?

File: src/Factory/ExampleGiftCardFactory.php (hunk 12 above)
Comment (Roshyo) [reply src/Factory/ExampleGiftCardFactory.php]: Because that way, it is clear that you have an ExampleGiftCard.
I think it is cleaner than having just a method in regular factory.

File: src/Resources/private/default-gift-card-configuration-css.txt (hunk 16 above)
Comment (Roshyo) [reply src/Resources/private/default-gift-card-configuration-css.txt]: To make it clear that it is no CSS loaded the regular way

File: src/Controller/Action/Admin/GenerateExamplePdfAction.php [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+use Setono\SyliusGiftCardPlugin\Form\Type\GiftCardConfigurationType;
+use Setono\SyliusGiftCardPlugin\Generator\GiftCardPdfGeneratorInterface;
+use Setono\SyliusGiftCardPlugin\Model\GiftCardConfigurationInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Webmozart\Assert\Assert;
+
+final class GenerateExamplePdfAction
```
Comment (loevgaard) [inline src/Controller/Action/Admin/GenerateExamplePdfAction.php]: So either this action should return the PDF and we will encode it client side (with https://developer.mozilla.org/en-US/docs/Web/API/btoa) or this action should be renamed to `GenerateEncodedExamplePdfAction` I think. Wdyt?

File: src/Resources/private/default-gift-card-configuration-css.txt (hunk 16 above)
Comment (loevgaard) [reply src/Resources/private/default-gift-card-configuration-css.txt]: Should this file be a `.css.twig` file and be rendered using twig? Wdyt?

File: src/Factory/GiftCardFactoryInterface.php [hunk 18]
```diff
@@ -24,4 +24,6 @@ public function createFromOrderItemUnitAndCart(
         OrderItemUnitInterface $orderItemUnit,
         OrderInterface $cart
     ): GiftCardInterface;
+
```
Comment (loevgaard) [inline src/Factory/GiftCardFactoryInterface.php:27]: Maybe add a comment here about what it's used for ;)

File: src/Resources/private/default-gift-card-configuration-css.txt (hunk 16 above)
Comment (Roshyo) [reply src/Resources/private/default-gift-card-configuration-css.txt]: Absolutely. And it simplifies some parts also :)

**Review by loevgaard — APPROVED** [review]
